### PR TITLE
Use files from ICU to implement UTF8 and UTF16 decoding

### PR DESCRIPTION
**Pull request by maxbrunsfeld**
This is a suggested change to https://github.com/tree-sitter/tree-sitter/pull/436. I thought I'd propose it via a PR since I had to explore it a little myself in order to determine if it worked well, but I didn't want to just apply the change without talking it through.

https://github.com/tree-sitter/tree-sitter/pull/436 is a nice PR. I'm glad to learn that `libicu` has a more performant unicode decoding routine than `utf8proc`. I also think that for dependencies like this, which are very small and need updating very infrequently, it's nice to avoid a git submodule, and to just include the necessary code in the repo instead. Lastly, removing `utf8proc.c` is good for compilation time. All of the unused code would eventually get removed from the final executable (or WASM module) anyway, but this lets us produce a smaller intermediate static library.

However, I'm not fully comfortable depending on a completely separate implementation of UTF8 decoding; I would rather use the *exact* same code that `libicu` uses. That way, we can be sure that no bugs have been introduced in porting the code, and if we ever want to incorporate further changes to `libicu`, it will be straightforward to pull them in.

What I've done in this PR is to drop the custom UTF8 decoding routine. While I was here, I also dropped the existing UTF16 decoding routine as well. In places of these files, I've copied 3 source files from `libicu` into the repository (along with the `libicu` license file).

The three main files that we need are `utf8.h` and `utf16.h` (which define the decoding routines) and `utf8machine.h` (which defines some basic types and constants). These three files are not very big (~ 800 lines each). These files reference some *other* source files via `#include`, but those source files contain functionality that isn't really needed, so I've added some empty files to allow those `#include` directives to resolve successfully.

### CODE REVIEW

File: lib/src/lexer.c [hunk 1]
```diff
@@ -54,16 +51,15 @@ static void ts_lexer__get_lookahead(Lexer *self) {
 
   // If this chunk ended in the middle of a multi-byte character,
   // try again with a fresh chunk.
-  if (self->lookahead_size == (uint32_t)DECODE_NEXT_ERROR && size < 4) {
+  if (self->data.lookahead == TS_DECODE_ERROR && size < 4) {
```
Comment (maxbrunsfeld) [inline lib/src/lexer.c:54]: This is a return to the approach on master of using the lookahead character (as opposed to the number of bytes consumed) to check for an error. The `U8_NEXT` and `U16_NEXT` macros explicitly set the code point (which is a signed integer) to `-1` aka `U_SENTINEL` on failure.

**Comment by mkrupcale** [comment]
> [tree-sitter#436](https://github.com/tree-sitter/tree-sitter/pull/436) is a nice PR.

I'm glad to hear you're amenable to the idea :).

> I also think that for dependencies like this, which are very small and need updating very infrequently, it's nice to avoid a git submodule, and to just include the necessary code in the repo instead.

Yes, those were my thoughts as well, considering how small a UTF-8 and UTF-16 decoder actually is.
 
> However, I'm not fully comfortable depending on a completely separate implementation of UTF8 decoding; I would rather use the _exact_ same code that `libicu` uses. That way, we can be sure that no bugs have been introduced in porting the code, and if we ever want to incorporate further changes to `libicu`, it will be straightforward to pull them in.

I did have this general concern as well when I was implementing my own version of the ICU implementation, although I should point out that I have validated my implementation against both invalid input sequences and the full set of possible valid code unit sequences[1]. Furthermore, as you pointed out above, the implementation of the UTF-8 and UTF-16 decoders is expected to change very infrequently, so it's unlikely this would have to be verified often (if at all--there's no real requirement to track ICU exactly at all times, given that the current decoder has been verified). In particular, the last notable change to `U8_NEXT` was `56fd8a74c5ee81026a9a2e2ffeb852c386d8f38b` on 2017-12-05, and the last notable change to `U16_NEXT` was `8bb10245f2ad6bab6ba2b1155d8af0d04425f0fe` on 2012-11-10.

> What I've done in this PR is to drop the custom UTF8 decoding routine...

I general, I'm not opposed to these changes, but I'd like to point out some drawbacks of this approach:

1. One of the things I did when I made the `libutf8decoders` project (which is mostly a toy project for benchmarking and testing decoders), was to try to come up with a somewhat standard API[2]. In particular, for the `decode_next` API[3], which is suited nicely for `tree-sitter`, I sought to keep the `code_point` return value separate from error-handling. I thought that the previous `utf8proc` API which somewhat conflated the two was confusing. Obviously, as long as you know how the API works and you use it correctly, it doesn't particularly matter, but I (obviously) preferred and felt the API I specified was cleaner and less error-prone.
  Now, this API could still be implemented using your approach of including the ICU headers directly, since you also are wrapping the macros in the `ts_decode_utf{8,16}` functions, but that will require checking for `U_SENTINEL` in these wrapper functions, returning `U_SENTINEL` in this case, and otherwise returning `i * n`. So you'd essentially be doing this test twice: once in the wrapper function, and once when calling the wrapper function. This is one of the reasons that I wrote my own version of the ICU implementation: so that I could get it to behave ideally according to my desired API, where instead of setting `code_point` to some error value, I return an error value directly.
2. There is a lot more code to add: this PR adds a net 2300 lines on top of my PR, which is a net of roughly 100 lines. This is essentially an order of magnitude difference. As you pointed out above, removing the `utf8proc` dependency reduces compile time. `utf8proc` is ~16 kloc, and this PR will add about 2 kloc, for a net loss of ~14 kloc, as opposed to the existing PR with a net loss closer to the full ~16 kloc. The rest of `tree-sitter` library is ~9 kloc, so this ~2 kloc difference could be a noticeable difference in compile time, now that `utf8proc` is removed.
3. Updates to the decoders still require manual intervention. Although most of this is now a purely verbatim copy, and you have nicely documented how this can be done, the process is still a manual one. As outlined above, though, these changes are expected to be very infrequent, regardless of how the update is done (if at all).
4. Given the size and nature of this change, I'd probably want to re-run the benchmark just to be sure nothing subtle has resulted in worse performance. This shouldn't be a too much of a problem (my current Criterion config takes about an hour to run), but still something to consider.

All in all, I'd prefer my approach, but let me know what you think.

[1] https://gitlab.com/mkrupcale/libutf8decoders#tests
[2] https://gitlab.com/mkrupcale/libutf8decoders#api
[3] https://gitlab.com/mkrupcale/libutf8decoders#decode_next

File: lib/src/lexer.c (hunk 1 above)
Comment (mkrupcale) [reply lib/src/lexer.c:54]: Yes, I noticed this and have commented somewhat on it below.

**Comment by mkrupcale** [comment]
Oh yes, I forgot also that the `U8_NEXT` and `U16_NEXT` macros do not actually handle the case where `length == 0`: they'll just read the first code unit regardless.

**Comment by maxbrunsfeld** [comment]
It does add more total code to the repo, but from a maintenance/risk standpoint, I'm not very concerned about the code being added - it's just a direct copy of a canonical implementation.

Regarding the error API, I prefer the error code going in the `code_point`, because it's convenient for the byte count to be an unsigned value, and code points seem to be conventionally signed (by both ICU and utf8proc). You can see in the diff that there's now less casting involved in using the decode functions.

**Comment by maxbrunsfeld** [comment]
> Oh yes, I forgot also that the U8_NEXT and U16_NEXT macros do not actually handle the case where length == 0: they'll just read the first code unit regardless.

Oh nice, thank you for catching that.

I'm just noticing now that you reordered things a bit in `ts_lexer__get_lookahead` so that it doesn't check the size before calling the function. I think I prefer the previous order: the only reason that the decode functions should return zero is if an error occurred (which we're already checking) or the buffer length was zero.

**Comment by mkrupcale** [comment]
> Regarding the error API, I prefer the error code going in the `code_point`, because it's convenient for the byte count to be an unsigned value, and code points seem to be conventionally signed (by both ICU and utf8proc).

Yes, I thought about this as well as I was designing the API, and I decided to use an unsigned code-point value partially because there are many bitwise manipulations done on this value, and if the code-point is at any point negative, either undefined or implementation-defined behavior can result. The down-side of using a signed return value for both "number of code units" or "error" I felt was outweighed by having a code-point behave well and be strictly nonnegative (given that code points cannot be negative anyways). Long term, I think efforts like [1] are the solution to this ambiguity in the return value.

> I'm just noticing now that you reordered things a bit in `ts_lexer__get_lookahead` so that it doesn't check the size before calling the function.

Yes, that's right, the check for length is now built into the `decode_next` API. I prefer this because it's a safer API to use, not making assumptions about the buffer even existing at all let alone having a nonzero size. Additionally, there is an inconsistency with respect to the `U8_NEXT` macros checking the length only partially (there is an implicit assumption of `length >= 1`).

> the only reason that the decode functions should return zero is if an error occurred (which we're already checking)

This case now returns `-1` to indicate error, which I personally think makes more sense than zero indicating an error.

> or the buffer length was zero.

This is satisfied by the current API: if the length of the buffer is zero, the function will return zero.

[1] http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1095r0.pdf#section.2

**Comment by mkrupcale** [comment]
Have you given any additional thought to the above points?

I'd like to reiterate the reasons for the choices I made and why I favor them over this PR:

1. The API design I chose is safer and easier to use:
   - Bounds are checked within the function
   - Error condition only requires checking the return value (not the output `code_point` parameter)
   - Unsigned integers are used for bitwise manipulation, as they should be
2. My PR results in a smaller code base:
   - Less code to maintain
   - Faster compiles with ~ 2 / 9 kloc difference
3. Using the exact ICU implementation is unnecessary:
   - My implementation has been tested and verified
   - Decoder updates should be very rare if ever
4. I need to re-verify this PR's current restructuring does not alter performance negatively

In light of the above, I propose three ways forward, in order of my preference:

1. We use my PR in which I make some small changes to variable and symbol names, prefixing `ts_`/`TS_` where appropriate, in line with this PR. This has the advantages listed above.
2. We use this PR with some modifications to keep the API I designed, while still simply wrapping the ICU implementation directly[1]. This has the advantage of maintaining point 1, while still using the exact ICU implementation.
3. We use this PR as is. This does not have any of the above advantages but uses the exact ICU implementation.

Pathways 2 and 3 should be verified according to point 4.

[1] https://gitlab.com/mkrupcale/libutf8decoders/blob/master/libutf8decoders/icu/icu.hpp#L96

**Comment by maxbrunsfeld** [comment]
Here's what I would say:

* I do want to use libicu's implementation, rather than maintaining a separate one.

* I don't want to change the representation of the `lookahead` field on `TSLexer`. It is currently an `int32_t`, and every Tree-sitter parser [is compiled to expect an `int32_t`](https://github.com/tree-sitter/tree-sitter-python/blob/6ec867cdabb4f237d87893b8b8e92423ee6efb1a/src/tree_sitter/parser.h#L43).

   I agree that in principle, using an unsigned value seems reasonable, but I don't want to break binary compatibility. Also, *both* `utf8proc` and `libicu` have set a convention that unicode code points are 32-bit signed integers, so I think it is also totally fine to abide by that convention.

> Bounds are checked within the function

* I don't want to make this change either. This is a low-level routine which is only called in one place in the codebase. And the calling function (`ts_lexer__get_lookahead`) already needs to check for EOF. When you added the bounds check, you had to change the caller in order to avoid *duplicate* bounds check, but I think the *original* place for the bounds check communicated the intent more directly.

> I need to re-verify this PR's current restructuring does not alter performance negatively

I would say go for it if you are interested, but I don't think it matters to me, due to the extremely small changes that we're talking about.

The normal benchmarks show no performance difference, and at the percentages we're talking about, your Criterion benchmarks are probably picking up a great deal of random fluctuations on your computer, and especially [unpredictable variations in code placement by the C compiler](https://emeryberger.com/research/stabilizer/) that have nothing to do with the code itself.

If you do conclude though that your implementation is faster than libicu's, it might be cool to submit a patch to libicu.

**Comment by mkrupcale** [comment]
> - I do want to use libicu's implementation, rather than maintaining a separate one.

Okay.

> - I don't want to change the representation of the lookahead field on TSLexer. It is currently an int32_t, and every Tree-sitter parser is compiled to expect an int32_t.

My PR does not propose to change this. In my PR, the signed `int32_t lookahead` type is only treated as an unsigned int during the decoding process, during which all of the bitwise manipulation occurs. This should work fine since its values should be well within range for a Unicode code point.

> I agree that in principle, using an unsigned value seems reasonable, but I don't want to break binary compatibility.

Understandable, and I don't intend to change that at this point.

> Also, both utf8proc and libicu have set a convention that unicode code points are 32-bit signed integers, so I think it is also totally fine to abide by that convention.

I would not use that fact to indicate that there is some sort of standard on code-points being represented using 32-bit signed integers. On the contrary, most of the implementations I tested used unsigned integers: BH[1], `Boost.Text`[2], branchless[3], `std::codecvt`[4], KEWB[5], LLVM[6] and Python[7,8].

> - And the calling function (`ts_lexer__get_lookahead`) already needs to check for EOF. When you added the bounds check, you had to change the caller in order to avoid duplicate bounds check, but I think the original place for the bounds check communicated the intent more directly.

I disagree because it is very much the concern of the decoder how many code units are requested for decoding as opposed to the calling function. Right now, the raw ICU implementation has an implicit assumption that `length >= 1`, despite what `length` is passed. This is an inconsistent assumption in which bounds are only partially checked by the decoder.

What the calling function decides to do then should be the result of what the decoder yields, including the EOF and error-handling. That is, the decoder should always be called to let it do the decoding safely. Regardless of how low-level this routine is, the safer API makes more sense from both a consistency and a separation-of-concerns standpoint.

> The normal benchmarks show no performance difference

Yes, this is why I implemented the Criterion benchmarks.

> and at the percentages we're talking about, your Criterion benchmarks are probably picking up a great deal of random fluctuations on your computer

I was careful to avoid this during my benchmarking. I was not using my workstation but rather a server whose hardware I control and of which I was the only user during benchmarking. To compare the different versions, I built from scratch both versions (i.e. cleaned the old build artifacts). The benchmarks were the only thing running during this time, and the results were reproducible and statistically significant.

Thus, I think that these differences are indeed real and not just random fluctuations.

> If you do conclude though that your implementation is faster than libicu's, it might be cool to submit a patch to libicu.

I was not concerned with the performance change of my decoder implementation compared to the ICU implementation, because they are essentially identical with the exception of the bounds check. The difference comes from how it is used in the TreeSitter caller as a result of this movement of the bounds check as well as the error-checking. It is a subtle difference that requires attention.

Having said all of this, I believe at this point we are trying to decide whether or not to follow proposal 2 or 3 above, where I still favor proposal 2. Alternatively, there's a proposal somewhere in between where the bounds-check is done outside of the decoder, but the error-checking is done through the return parameter, and `uint32_t` is used for the code-point.

[1] http://bjoern.hoehrmann.de/utf-8/decoder/dfa/
[2] https://tzlaine.github.io/text/doc/html/boost/text/v1/utf_8_to_32_iterator.html#idm25874-bb
[3] https://github.com/skeeto/branchless-utf8/blob/f2d0e24c3864d726cd009901726df4778ad3e0d5/utf8.h#L26
[4] https://en.cppreference.com/w/cpp/locale/codecvt
[5] https://github.com/BobSteagall/CppNow2018/blob/dcf5f945449ec38fc4864a597ed6c87511bce9cf/FastConversionFromUTF-8/code/src/utf_utils.h#L107
[6] https://llvm.org/doxygen/ConvertUTF_8h_source.html#l00109
[7] https://docs.python.org/3/c-api/unicode.html
[8] https://github.com/python/cpython/blob/6876257eaabdb30f27ebcbd7d2557278ce2e5705/Include/unicodeobject.h#L102

**Comment by maxbrunsfeld** [comment]
I don't want to use `uint32_t` to represent the code-point, because ICU returns a signed integer, and the struct field `lookahead` is a signed integer, so using an unsigned integer for the intermediate step just introduces unnecessary type casts.

> I disagree because it is very much the concern of the decoder how many code units are requested for decoding as opposed to the calling function.

It is the concern of both functions. The caller performs different logic at EOF. It's a very minor detail, but I prefer that the caller perform the initial length check.

> I was careful to avoid this during my benchmarking. I was not using my workstation but rather a server whose hardware I control

Ok but just curious - is there an operating system running on that computer? Did you repeatedly alternate between benchmarking the old code and the new code, or did you just benchmark one version for an extended period of time, followed by the other version? If the latter, then your results are very vulnerable to skew due to changes in the operating system's workload over time.

Even beyond that, we aren't using a code-layout-randomization tool like [`Stabilizer`](https://emeryberger.com/research/stabilizer/) (see [this talk](https://www.youtube.com/watch?v=r-TLSBdHe1A)), so there is a huge amount of randomness in any performance measurement, which may or may match what happens in downstream applications. Takeaway - there are very-quickly-diminishing returns to increasing statistical sophistication in benchmarking.

**Comment by mkrupcale** [comment]
> ICU returns a signed integer

This is not strictly correct, because `U8_NEXT` is technically a macro which makes no explicit assumptions on the type of the code point, `c`[1]. The type will come from the wrapper function `ts_decode_utf8`.

> so using an unsigned integer for the intermediate step just introduces unnecessary type casts.

Sure, but in avoiding the type casts one must do bitwise manipulations on the signed integer, which can invoke undefined or implementation-defined behavior. I suppose that as long as the code point stays within the Unicode range (i.e. non-negative), this should be relatively safe, however.

> Ok but just curious - is there an operating system running on that computer?

I'm sure you're being facetious here, but yes, I was using an operating system. The point being, however, that this benchmark was the primary task being run.

> Did you repeatedly alternate between benchmarking the old code and the new code, or did you just benchmark one version for an extended period of time, followed by the other version?

On two completely separate occassions, months separated, I tested these changes and observed consistent relative changes of similar statistical significance (i.e. the number of benchmarks in the "regression", "unchanged or within the noise", and "improvement" categories were similar). The benchmarks themselves were run using Criterion over a period of ~40 minutes to ~4 hours (for both original and modified code). Criterion has its own procedure for warm-up between benchmarks.

> If the latter, then your results are very vulnerable to skew due to changes in the operating system's workload over time.

As I said, I was the only person using the computer during this time, and this was the only "significant" task running during that time.

> Even beyond that, we aren't using a code-layout-randomization tool like Stabilizer (see this talk), so there is a huge amount of randomness in any performance measurement, which may or may match what happens in downstream applications.

Certainly there are many factors at play when it comes to measuring such performance differences. The effects he is talking about are random layout in function addresses, stack frame sizes, and heap allocations. I'm not disputing that these things can affect your performance measurement. However, I'm not entirely convinced that these are "random" effects. As far as I'm aware, these are deterministic for a given toolchain executed sequentially. Even for PIC/PIE with ASLR, unless he is trying to argue that absolute positioning of the base address is significant.

Thus, he is trying to measure variation between toolchains by introducing randomness into the toolchain. However, what I'm arguing is that under the conditions that I tested--which were such that I reduced to the extent possible all unwanted variations--for a given toolchain, I do observe a statistically significant improvement on average across all of the benchmarks.

> Takeaway - there are very-quickly-diminishing returns to increasing statistical sophistication in benchmarking.

Actually, I believe that is essentially the opposite of the conclusions of the talk and my own views. Rather, I would say that reliably benchmarking can be challenging and requires careful statistical analysis. Otherwise, there would essentially be no hope of detecting "real" performance changes, and we ought not even bother measuring them at all.

Okay, so I think at this point we have settled on:
- Using the exact ICU decoder implementations
- Using a wrapper function which passes an `int32_t` for the code point

Yet to be conclusively decided are:
- `size == 0` bound check: 1) inside the decoder wrapper; or 2) in the calling function
- error return channel: 1) return value; or 2) code point output parameter

Aside: related to the error return channel, this may be interesting for future C2x: https://www.reddit.com/r/cpp/comments/dc8kse/wg14_function_failure_annotation_combined/

[1] https://github.com/unicode-org/icu/blob/64b58ccda315902b7b88e32a12ffca3b7b99cdf0/icu4c/source/common/unicode/utf8.h#L380

**Comment by maxbrunsfeld** [comment]
> I'm not entirely convinced that these are "random" effects. As far as I'm aware, these are deterministic for a given toolchain executed sequentially.

You're right; they're not literally random. What I means is that they're not predictable, and they are very likely to be different in the benchmark than they do in any final application *using* Tree-sitter.

> This is not strictly correct, because U8_NEXT is technically a macro which makes no explicit assumptions on the type of the code point, c[1].

No, it does explicitly state the signedness. On failure, it returns `U_SENTINEL`, which is `-1`, and the documentation states `@param c output UChar32 variable, set to < 0 in case of an error`.

> `size == 0` bound check: 1) inside the decoder wrapper; or 2) in the calling function
> * error return channel: 1) return value; or 2) code point output parameter

My preference is as it stands in this PR.

**Comment by mkrupcale** [comment]
> You're right; they're not literally random. What I means is that they're not predictable, and they are very likely to be different in the benchmark than they do in any final application using Tree-sitter.

Even with this interpretation of randomness, however, he doesn't exactly argue for "how much" randomness (i.e. statistical variance) ought to be introduced to represent the variations you could see in actual toolchains. More accurate then I believe would be to perform the measurements with different toolchains and platforms and compare. However, it's somewhat unreasonable to expect to capture all such variations, so instead I only claim to have observed statistically significant improvments on the platform that I measured. This is about the best I think one could hope for for a change such as this. I would be interested to see if others observe different results, but that would require them to take the time to try it for themselves.

> On failure, it returns `U_SENTINEL`, which is `-1`

This is a signed literal, yes, but it will be converted to whatever type `c` is.

> and the documentation states `@param c output UChar32 variable, set to < 0 in case of an error`.

Yes, this is what the documentation says, but just reading the macro, you will see that no such actual constraint is placed on the type of `c`.

> My preference is as it stands in this PR.

I'll do some benchmarking on this PR and let you know what I find.

**Comment by mkrupcale** [comment]
### Runtime performance comparisons

I ran Criterion benchmarks with both the `tree-sitter/utf8proc-remove` and `mkrupcale/utf8proc-remove` branches on top of `master` commit `20bb99249bafc36d6e4c0e39af86fae4c18577a2`. I ran the Criterion benchmarks with `N=20`, warm-up time of `500 ms`, measurement time of at least `1 s`, and `10^5` resamples. Each commit/branch was benchmarked in this way twice to verify reproducibility.

#### `tree-sitter/utf8proc-remove`

Number of benchmark (parser, input) combinations falling into the categories (out of 455 total):

* Regressions: [238, 223] ~ 231
* Unchanged or within noise: [136, 145] ~  140
* Improvements: [81, 87] ~ 84

Average relative times (relative to `master`, lower is better):
[1.00435, 1.00568] ~ 1.005015

#### `mkrupcale/utf8proc-remove`

Number of benchmark (parser, input) combinations falling into the categories (out of 455 total):

* Regressions: [133, 189] ~ 161
* Unchanged or within noise: [223, 164] ~ 193
* Improvements: [99, 102] ~ 101

Average relative times (relative to `master`, lower is better):
[0.997442, 1.00045] ~ 0.998946

### Compile time performance comparisons

Comile time of the tree-sitter library was measured using `script/build-lib` iterated 100 times.

#### `tree-sitter/master`

```shell
real    9m0.408s
user    8m48.183s
sys     0m5.598s
```

#### `tree-sitter/utf8proc-remove`

```shell
real    8m48.297s
user    8m37.528s
sys     0m4.219s
```

Relative time (compared to `master`): ~0.98

#### `mkrupcale/utf8proc-remove`

```shell
real    7m45.956s
user    7m35.699s
sys     0m3.881s
```

Relative time (compared to `master`): ~0.86

### Conclusions

The runtime performance comparisons show that there is probably a slightly better improvement (and fewer regressions) using `mkrupcale/utf8proc-remove` than with `tree-sitter/utf8proc-remove`, resulting in relative average time improvement of ~0.1% versus ~0.5% regression. The difference between these branches is quite subtle as I indicated previously and has primarily to do with the structure of `ts_lexer__get_lookahead`.

Note that these performance comparisons are not as favorable (cf. ~2% improvement) as when I benchmarked a couple months ago. It's possible that there have been some improvements on `master` since then, diminishing the effect of the decoder improvements.

There is a ~14% reduction in compile times with the `mkrupcale/utf8proc-remove` branch relative to `master` versus only a ~2% reduction for `tree-sitter/utf8proc-remove`. This represents a ~12% reduction in compile times of `mkrupcale/utf8proc-remove` relative to `tree-sitter/utf8proc-remove`.

If these metrics are satisfactory, I'll go ahead and accept this PR, but `lib/src/query.c` requires a patch to compile:

[query.c-utf8-decoder.txt](https://github.com/mkrupcale/tree-sitter/files/3721870/query.c-utf8-decoder.txt)

### Benchmark logs

[criterion-utf8proc-remove.tar.gz](https://github.com/mkrupcale/tree-sitter/files/3721795/criterion-utf8proc-remove.tar.gz)

**Comment by maxbrunsfeld** [comment]
Interesting! I do think the results from this branch are satisfactory. Thanks for taking the time to report all of that.

**Comment by mkrupcale** [comment]
No problem, do you want to go ahead and push the patch for `query.c` I attached first before I merge this?

**Comment by maxbrunsfeld** [comment]
I would say you could just merge this PR as-is into your branch, and then when I merge your PR into master, I will make sure to include that patch to `query.c`. Thanks for calling that out. ⚡️

**Comment by mkrupcale** [comment]
Sounds good. Done.

**Pull request merged**